projects/ngx-datatable/src/lib/directives/draggable.directive.ts

```diff
@@ -1,151 +1,138 @@
+    if (!this.enabled()) {
       return;
     }
-
-    const { clientX, clientY } = getPositionFromEvent(event);
-    const x = clientX - mouseDownPos.clientX;
-    const y = clientY - mouseDownPos.clientY;
-
-    this.last = { x, y };
-
-    this.frameId ??= requestAnimationFrame(() => this.updatePosition());
-
-    this.dragging.emit({
-      event,
-      element: this.element,
-      model: this.dragModel()
+    event.stopPropagation();
+    event.preventDefault();
+    const touch = event.touches.item(0)!;
+    this.touchId = touch.identifier;
+
+    const moveListener = (moveEvent: TouchEvent) => {
+      const { clientX, clientY } = this.findTouch(moveEvent)!;
+      this.moving(clientX, clientY);
+    };
+    const endListener = () => {
+      this.document.removeEventListener('touchmove', moveListener);
+      this.document.removeEventListener('touchend', endListener);
+      this.ending();
+    };
+
+    this.document.addEventListener('touchend', moveListener);
+    this.delay(this.dragStartDelay()).then(() => {
+      this.startX.set(touch.clientX);
+      this.startY.set(touch.clientY);
+      this.document.addEventListener('touchmove', moveListener);
+      this.starting();
     });
   }
 
-  onMouseup(event: MouseEvent | TouchEvent): void {
-    if (!this.isDragging()) {
-      return;
-    }
-
-    this.isDragging.set(false);
-    this._destroySubscription();
-    this.dragEnd.emit({
-      event,
-      element: this.element,
-      model: this.dragModel()
+  protected starting(): void {
+    this.dragStart.emit();
+  }
+  protected moving(clientX: number, clientY: number): void {
+    this.dragMove.emit({
+      initialX: this.startX()!,
+      initialY: this.startY()!,
+      currentX: clientX,
+      currentY: clientY
     });
   }
 
-  private updatePosition(): void {
-    this.frameId = undefined;
-    if (!this.last) {
-      return;
-    }
-    if (this.dragX()) {
-      this.element.style.left = `${this.last.x}px`;
-    }
-    if (this.dragY()) {
-      this.element.style.top = `${this.last.y}px`;
+  protected ending(): void {
+    const dragged = this.isDragging();
+    this.touchId = undefined;
+    this.startX.set(undefined);
+    this.startY.set(undefined);
+    clearTimeout(this.timeoutId);
+    // This function is also called if the long press was aborted before the delay.
+    // In that case, we don't want to emit dragEnd.
+    if (dragged) {
+      this.dragEnd.emit();
     }
   }
 
-  private _destroySubscription(): void {
-    if (this.frameId) {
-      cancelAnimationFrame(this.frameId);
-    }
+  private findTouch(event: TouchEvent): Touch | undefined {
+    return Array.from(event.touches).find(touch => touch.identifier === this.touchId);
+  }
 
-    if (this.subscription) {
-      this.subscription.unsubscribe();
-      this.subscription = undefined;
-    }
+  private delay(ms: number): Promise<void> {
+    return new Promise(resolve => (this.timeoutId = window.setTimeout(() => resolve(), ms)));
```

Comment:
this can leak if component is destroyed before `ending` method being called by the listener. Why not using rxjs `timer` here?

Reply:
Clearing the timeout in ondestroy. No need for rxjs here where anyway also would need to unsubscribe...